11x-new-kafka-roller.md

```diff
@@ -0,0 +1,263 @@
+Also, in the current roller, if we failed to reconfigure a node, we immediately restart it.
+Reconfiguration can fail sometimes due to transitive error so it would be useful to retry the reconfiguration a few times before we decide to restart a node.
+
+In each reconciliation, the number of attempts is tracked for each node.
+The number of attempts is how many times the overall process is repeated per node because of not reaching the desired state and the number of restarts is how many times a node is actually restarted.
+If any node has reached the maximum number of attempts or restarts, the reconciliation will fail.
+If the maximum number of reconfiguration is reached, then the node will be marked to restart but will not fail the reconciliation.
+When a new reconciliation starts, the tracked number of actions taken on nodes will be reset.
+
+The current roller also fails the reconciliation in the following situation to avoid continuing the process on more nodes:
+- Cannot connect to KafkaAgent to check broker state metrics (KafkaAgent is crucial when determining safety before rolling nodes, so there is no point to try to progress further, if we cannot connect to the KafkaAgent).
+- Pod is not Running and does not have an old revision (this will prevent the roller from restarting more nodes, which could result in making them stuck as well and bring down the entire cluster).
+
+### Batch rolling
+
+Batch rolling is one of the major features that the new roller is introducing.
+The proposed algorithm is to group broker nodes without common partitions together for parallel restart while maintaining availability.
+
+The proposed algorithm does not take rack information into account and the reason for this is explained in the `Rejected Alternatives` section.
+When batching brokers without rack awareness is that batch size would likely decrease as the roller progresses and potentially result in brokers that cannot be batched together so they would be restarted one by one.
+However, the majority of brokers would likely to get restarted in parallel and that would still speed the rolling in large clusters significantly.
+
+There is also an interesting future improvement that can make the batch rolling more effective optimizing with Cruise Control.
+This improvement will not be in the scope of this proposal but included in `Future Improvements` section.
+
+The batching algorithm only applies to broker nodes, however, the capability to roll nodes in parallel will be used for controller nodes as well.
+This is needed when a controller quorum is in a bad state.
+[#9426](https://github.com/strimzi/strimzi-kafka-operator/issues/9426) mentioned in the `Known Issues` section, is an example of why this feature is important for controller nodes.
+The new roller will check if there are multiple controller nodes not working affecting the quorum, and restart them in parallel to help recover it.
```

Comment:
```suggestion
The new roller will check if there are multiple controller nodes not working affecting the quorum, and restart them in parallel to help recover it.
Unlike the batch rolling of brokers, the user cannot disable the roller from batch rolling the controllers.
This is because the batch rolling will only be used in the case where multiple controllers are unavailable, affecting the quorum.
```